### PR TITLE
[WIP] Introduces tenant functionality for Cache operations

### PR DESCRIPTION
**Pull request by vbekiaris**
Based on & supersedes #10874 , introduces ability to define a `TenantControl` that sets up the context for execution of Cache operations. `TenantControl` is loaded via `ServiceLoader` mechanism (instead of `Config` in #10874 ).

### CODE REVIEW

**Comment by mmedenjak** [comment]
I am sorry that I have to do this but unfortunately we're again at the end of the release. Moving to 3.12. I hope this will be one of the first PRs to get merged into 3.12 though.

**Comment by lprimak** [comment]
Since this PR has been completely rewritten from the original PR, I cannot test it just by looking at the code.
However, since it doesn't break anything please merge it in

File: hazelcast/src/main/java/com/hazelcast/cache/impl/tenantcontrol/CacheDestroyEventContext.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    public CacheDestroyEventContext() {
+    }
+
+    public CacheDestroyEventContext(String cacheDistributedObjectName) {
+        this.cacheDistributedObjectName = cacheDistributedObjectName;
+    }
+
+    @Override
+    public void destroy(ICacheService context) {
+        context.destroyDistributedObject(getDistributedObjectName(cacheDistributedObjectName));
```
Comment (lprimak) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/tenantcontrol/CacheDestroyEventContext.java]: Does this destroy the cache?  Or just remove local node configuration?
The concept behind this is to "disconnect" local node from the cache, not to remove it entirely.

Comment (lprimak) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/tenantcontrol/CacheDestroyEventContext.java]: See issue #10875 for explanation
thanks

**Comment by vbekiaris** [comment]
Closing temporarily - will reopen once conflicts are resolved and ready for review

**Comment by vbekiaris** [comment]
Reopened in #14633

**Pull request closed without merging**